documentation/gnmi/dash-gnmi-design.md

```diff
@@ -0,0 +1,133 @@
+# GNMI Interface for DASH
+
+## Overview
+
+SONiC should provide a gNMI server interface for DASH project.
+Below diagram shows the architecture of the DASH project, the host server and DPU card are connected through PCIE, and together they form a smart switch.
+
+![dash-arch](./images/gnmi-arch.svg)
+
+Both host server and DPU card are running SONiC image. GNMI server is on host server, orchagent and redis are on DPU card, GNMI server will use ZMQ over TCP to communicate with orchagent and redis.
+
+### Requirements
+
+* Set and get RPCs must be supported. Customers will use get RPC to retrieve DASH configurations and use set PRC to apply new DASH configurations.
+* Need to configure huge DASH table entries to APPL_DB, with high speed.
+* Minimal redis table scaling requirements:
+
+|Item |Expected value |
+|--|--|
+|VNETs|1024|
+|ENI|64 per card|
+|Outbound Routes per ENI|100K|
+|Inbound Routes per ENI|10K|
+|NSGs per ENI|10|
+|ACL rules per NSG|1000|
+|ACL prefixed per ENI|10*100K|
+|Max prefixed per rule|8K|
+|ACL ports per ENI|10*10K SRC/DST ports|
+|CA-PA mapping|10M per card|
+
+### Design considerations
+
+* For set RPC, GNMI calls the ZMQ interface, and ZMQ updates redis asynchronously.
+* For get RPC, GNMI reads from redis db directly.
+* The DASH table of APPL_DB would be encoded as protobuf to save memory consumption, and then GNMI needs to support protobuf encoding for DASH table.
```

Comment:
>encoded as protobuf [](http://example.com/codeflow?start=37&length=19)

Could you provide a sample of "encoded as plaintext" for comparison purpose?

Reply:
I add IETF JSON encoding message for comparison purpose.